### ISSUE TITLE
build: mingw GUIX build is failing on current master

### ISSUE THREAD

**Issue by laanwj** [bitcoin/bitcoin#24055]
Current HEAD commit `290ff5ef6d3886409e5e8688f7d28a4c8246c617` fails to build on guix (on a GUIX VM) here:

Build command
```bash
time ( SOURCES_PATH="${HOME}/sources" BASE_CACHE="${HOME}/base-cache" SDK_PATH="${HOME}/sdk-path" contrib/guix/guix-build ) 2>&1 | tee "${HOME}/build.log"
```

Last output:
```
INFO: Building 290ff5ef6d38 for platform triple x86_64-linux-gnu:
      ...using reference timestamp: 1642008478
      ...running at most 8 jobs
      ...from worktree directory: '/home/guest/bitcoin'
          ...bind-mounted in container to: '/bitcoin'
      ...in build directory: '/home/guest/bitcoin/guix-build-290ff5ef6d38/distsrc-290ff5ef6d38-x86_64-linux-gnu'
          ...bind-mounted in container to: '/distsrc-base/distsrc-290ff5ef6d38-x86_64-linux-gnu'
      ...outputting in: '/home/guest/bitcoin/guix-build-290ff5ef6d38/output/x86_64-linux-gnu'
          ...bind-mounted in container to: '/outdir-base/x86_64-linux-gnu'
Updating channel 'guix' from Git repository at 'https://git.savannah.gnu.org/git/guix.git'...
Computing Guix derivation for 'x86_64-linux'...  
substitute: updating substitutes from 'https://ci.guix.gnu.org'... 100.0%
substitute: updating substitutes from 'https://bordeaux.guix.gnu.org'... 100.0%
The following derivation will be built:
   /gnu/store/cscw4gwah43lagqlljb19s26dhvlr6m6-profile.drv

building package cache...
builder for `/gnu/store/ni374dsml56hdgsqdg2zpa5qvl1mban4-guix-package-cache.drv' failed to produce output path `/gnu/store/bd7zd6fx23k3z18c73zb288yz8v6f5lb-guix-package-cache'
build of /gnu/store/ni374dsml56hdgsqdg2zpa5qvl1mban4-guix-package-cache.drv failed
View build log at '/var/log/guix/drvs/ni/374dsml56hdgsqdg2zpa5qvl1mban4-guix-package-cache.drv.bz2'.
note: keeping build directory `/tmp/guix-build-guix-package-cache.drv-1'
cannot build derivation `/gnu/store/cscw4gwah43lagqlljb19s26dhvlr6m6-profile.drv': 1 dependencies couldn't be built
guix time-machine: error: build of `/gnu/store/cscw4gwah43lagqlljb19s26dhvlr6m6-profile.drv' failed
```

Looking in the mentioned log:
```
$ bunzip2 < /var/log/guix/drvs/ni/374dsml56hdgsqdg2zpa5qvl1mban4-guix-package-cache.drv.bz2
(repl-version 0 1 1)
Generating package cache for '/gnu/store/h9app05jj4bc45fjp1hgld86ij6n8bv4-profile'...
(exception unbound-variable (value #f) (value "Unbound variable: ~S") (value (gnupg-2.2.32)) (value #f))
```

**Comment by 0xB10C** [comment]
Can reproduce on 31db3dd874df (the next commit after 290ff5ef6d3886409e5e8688f7d28a4c8246c617). Guess it's not a problem with your setup.

**Comment by laanwj** [comment]
It looks like #24057 fixes this. At least the process continues longer, will take a while to finish.

**Comment by laanwj** [comment]
It fixed the original problem. However it fails in a new place while building the win32 cross-compiler now:
```
building /gnu/store/6qvzfb0sf24giqril64jsymspng6avyb-gcc-cross-x86_64-w64-mingw32-10.3.0.drv...
note: keeping build directory `/tmp/guix-build-gcc-cross-x86_64-w64-mingw32-10.3.0.drv-0'
builder for `/gnu/store/6qvzfb0sf24giqril64jsymspng6avyb-gcc-cross-x86_64-w64-mingw32-10.3.0.drv' failed with exit code 1
build of /gnu/store/6qvzfb0sf24giqril64jsymspng6avyb-gcc-cross-x86_64-w64-mingw32-10.3.0.drv failed
View build log at '/var/log/guix/drvs/6q/vzfb0sf24giqril64jsymspng6avyb-gcc-cross-x86_64-w64-mingw32-10.3.0.drv.bz2'.
building /gnu/store/1jqzirixzsmkcbzg3mk62h5l1cncgwrn-mingw-w64-x86_64-winpthreads-8.0.0.drv...
cannot build derivation `/gnu/store/jd9dfysdmk7y8hq2abdrllyy9hmns7l5-nsis-x86_64-3.05.drv': 1 dependencies couldn't be built
note: keeping build directory `/tmp/guix-build-mingw-w64-x86_64-winpthreads-8.0.0.drv-0'
cannot build derivation `/gnu/store/2i9lg9yxrbir7rhkda9zd8hjz3mnrawi-profile.drv': 1 dependencies couldn't be built
guix environment: error: build of `/gnu/store/2i9lg9yxrbir7rhkda9zd8hjz3mnrawi-profile.drv' failed
```
With a linker error:
```
/gnu/store/4y5m9lb8k3qkb1y9m02sw9w9a6hacd16-bash-minimal-5.1.8/bin/bash ../../../gcc-10.3.0/libgcc/../mkinstalldirs .
ln -s -f libgcc.map libgcc.map.def && if [ ! -d ./shlib ]; then mkdir ./shlib; else true; fi && /tmp/guix-build-gcc-cross-x86_64-w64-mingw32-10.3.0.drv-0/build/./gcc/xgcc -B/tmp/guix-build-gcc-cross-x86_64-w64-mingw32-10.3.0.drv-0/build/./gcc/ -B/gnu/store/4fh9s2n54m57gcapc0kslaicc4pxf13g-mingw-w64-x86_64-8.0.0/lib    -O2 -I../../../gcc-10.3.0/libgcc/../winsup/w32api/include -g -O2 -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE  -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wno-error=format-diag -Wstrict-prototypes -Wmissing-prototypes -Wno-error=format-diag -Wold-style-definition  -isystem ./include   -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector   -shared -nodefaultlibs libgcc.map.def -Wl,--out-implib,./shlib/libgcc_s.a.tmp -o ./shlib/libgcc_s_seh-1.dll.tmp -g -O2 -B./ _chkstk_s.o _chkstk_ms_s.o _muldi3_s.o _negdi2_s.o _lshrdi3_s.o _ashldi3_s.o _ashrdi3_s.o _cmpdi2_s.o _ucmpdi2_s.o _clear_cache_s.o _trampoline_s.o __main_s.o _absvsi2_s.o _absvdi2_s.o _addvsi3_s.o _addvdi3_s.o _subvsi3_s.o _subvdi3_s.o _mulvsi3_s.o _mulvdi3_s.o _negvsi2_s.o _negvdi2_s.o _ctors_s.o _ffssi2_s.o _ffsdi2_s.o _clz_s.o _clzsi2_s.o _clzdi2_s.o _ctzsi2_s.o _ctzdi2_s.o _popcount_tab_s.o _popcountsi2_s.o _popcountdi2_s.o _paritysi2_s.o _paritydi2_s.o _powisf2_s.o _powidf2_s.o _powixf2_s.o _powitf2_s.o _mulhc3_s.o _mulsc3_s.o _muldc3_s.o _mulxc3_s.o _multc3_s.o _divhc3_s.o _divsc3_s.o _divdc3_s.o _divxc3_s.o _divtc3_s.o _bswapsi2_s.o _bswapdi2_s.o _clrsbsi2_s.o _clrsbdi2_s.o _fixunssfsi_s.o _fixunsdfsi_s.o _fixunsxfsi_s.o _fixsfdi_s.o _fixdfdi_s.o _fixxfdi_s.o _fixunssfdi_s.o _fixunsdfdi_s.o _fixunsxfdi_s.o _floatdisf_s.o _floatdidf_s.o _floatdixf_s.o _floatundisf_s.o _floatundidf_s.o _floatundixf_s.o _divdi3_s.o _moddi3_s.o _divmoddi4_s.o _udivdi3_s.o _umoddi3_s.o _udivmoddi4_s.o _udiv_w_sdiv_s.o gthr-win32_s.o cpuinfo_s.o sfp-exceptions_s.o addtf3_s.o divtf3_s.o eqtf2_s.o getf2_s.o letf2_s.o multf3_s.o negtf2_s.o subtf3_s.o unordtf2_s.o fixtfsi_s.o fixunstfsi_s.o floatsitf_s.o floatunsitf_s.o fixtfdi_s.o fixunstfdi_s.o floatditf_s.o floatunditf_s.o fixtfti_s.o fixunstfti_s.o floattitf_s.o floatuntitf_s.o extendsftf2_s.o extenddftf2_s.o extendxftf2_s.o trunctfsf2_s.o trunctfdf2_s.o trunctfxf2_s.o enable-execute-stack_s.o unwind-seh_s.o unwind-sjlj_s.o unwind-c_s.o emutls_s.o libgcc.a  -lmingwthrd -lmingw32 -lmingwex -lmoldname -lmsvcrt -ladvapi32 -lshell32 -luser32 -lkernel32 && if [ -f ./shlib/libgcc_s_seh-1.dll ]; then mv -f ./shlib/libgcc_s_seh-1.dll ./shlib/libgcc_s_seh-1.dll.backup; else true; fi && mv ./shlib/libgcc_s_seh-1.dll.tmp ./shlib/libgcc_s_seh-1.dll && mv ./shlib/libgcc_s.a.tmp ./shlib/libgcc_s.a
x86_64-w64-mingw32-ld: final link failed: bad value
collect2: error: ld returned 1 exit statu
```

**Comment by laanwj** [comment]
I have this issue again with master as of 219d728fcbde8c313940788838afa46c2fb88762.
```
building /gnu/store/6qvzfb0sf24giqril64jsymspng6avyb-gcc-cross-x86_64-w64-mingw32-10.3.0.drv...
note: keeping build directory `/tmp/guix-build-gcc-cross-x86_64-w64-mingw32-10.3.0.drv-1'
builder for `/gnu/store/6qvzfb0sf24giqril64jsymspng6avyb-gcc-cross-x86_64-w64-mingw32-10.3.0.drv' failed with exit code 1
build of /gnu/store/6qvzfb0sf24giqril64jsymspng6avyb-gcc-cross-x86_64-w64-mingw32-10.3.0.drv failed
View build log at '/var/log/guix/drvs/6q/vzfb0sf24giqril64jsymspng6avyb-gcc-cross-x86_64-w64-mingw32-10.3.0.drv.bz2'.
cannot build derivation `/gnu/store/i2pp5in7bl4ghq8apphmwdjsksw724pr-nsis-x86_64-3.05.drv': 1 dependencies couldn't be built
guix environment: error: build of `/gnu/store/i2pp5in7bl4ghq8apphmwdjsksw724pr-nsis-x86_64-3.05.drv' failed
```

```
x86_64-w64-mingw32-ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:994: libgcc_s.dll] Error 1
make[2]: Leaving directory '/tmp/guix-build-gcc-cross-x86_64-w64-mingw32-10.3.0.drv-0/build/x86_64-w64-mingw32/libgcc'
make[1]: *** [Makefile:12748: all-target-libgcc] Error 2
make[1]: Leaving directory '/tmp/guix-build-gcc-cross-x86_64-w64-mingw32-10.3.0.drv-0/build'
make: *** [Makefile:957: all] Error 2
error: in phase 'build': uncaught exception:
%exception #<&invoke-error program: "make" arguments: ("-j" "8" "FLAGS_FOR_TARGET=-B/gnu/store/4fh9s2n54m57gcapc0kslaicc4pxf13g-mingw-w64-x86_64-8.0.0/lib" "CFLAGS=-g0 -O2") exit-status: 2 term-signal: #f stop-signal: #f>
phase `build' failed after 409.0 seconds
command "make" "-j" "8" "FLAGS_FOR_TARGET=-B/gnu/store/4fh9s2n54m57gcapc0kslaicc4pxf13g-mingw-w64-x86_64-8.0.0/lib" "CFLAGS=-g0 -O2" failed with status 2
```

**Comment by dongcarl** [comment]
Upstream links:
Issue: https://issues.guix.gnu.org/53426
Patch: https://issues.guix.gnu.org/53706

I'm also testing a branch that fixes this in `manifest.scm` without any upstream intervention (ofc fixed upstream is better).

**Comment by dongcarl** [comment]
The workaround is now a blocker for the next Guix release: https://debbugs.gnu.org/cgi/bugreport.cgi?bug=53214;msg=51

Fixing in `manifest.scm` is turning out to be a little more involved than I thought, but will keep at it.

**Comment by dongcarl** [comment]
Patch pushed to master for Guix: https://git.savannah.gnu.org/cgit/guix.git/commit/?id=2d78b27711d44f39fabf170d044b45247af7babe

I'd recommend that we wait until the next convergence (when the `guix` package within Guix is bumped) to bump our time-machine target. However, if 1.4.0 is released before our next release, we should target that instead.

**Comment by hebasto** [comment]
> Patch pushed to master for Guix: https://git.savannah.gnu.org/cgit/guix.git/commit/?id=2d78b27711d44f39fabf170d044b45247af7babe
> 
> I'd recommend that we wait until the next convergence (when the `guix` package within Guix is bumped) to bump our time-machine target. However, if 1.4.0 is released before our next release, we should target that instead.

Considering recently merged #20744 and its follow ups that really should be tested for Windows target, I'd suggest to switch to the current top commit in the Guix master branch, and unfreeze Windows builds.

**Comment by dongcarl** [comment]
Sounds good, feel free to PR!

**Comment by laanwj** [comment]
This is confirmed fixed by #24348.
